### ISSUE TITLE
Creating a backup while the grain is running could cause corruption.

### ISSUE THREAD

**Issue by zenhack** [sandstorm-io/sandstorm#3186]
@orblivion brought this up on the mailing list, and this is an issue I was aware of but I don't think we have an open ticket for it so here it is:

If we take a .zip backup while a grain is running, there is a possibility of data corruption of the grain is modifying the filesystem while we are copying it. We should shut down the grain while taking the backup to ensure integrity.

**Comment by xet7** [comment]
This depends how backup is implemented.

Currently, because backup is made with making .zip file directly from raw files, this corruption can happen.

If there would be backup functionality in app itself, like PHP/JS/Go etc accesses database with TCP, then corruption would be not happen. There could be some Sandstorm API that call that goes through each grain's currently running database with TCP and makes backup, like is done with mysqldump, mongodump, etc. That kind of backup is also much safer to restore to that app that is running at Sandstorm or outside of Sandstorm. For example, [export Wekan grain to .zip file and move to standalone Wekan](https://github.com/wekan/wekan/wiki/Export-from-Wekan-Sandstorm-grain-.zip-file) instructions are so long because it is dealing with raw database files, and there is not yet feature in Wekan to export all boards, users, etc everything in .zip file.

**Comment by zenhack** [comment]
Right, I was thinking about this in terms of "there is a bug in the
current feature," but you could definitely also just look at other ways
of doing backups entirely.

It would be nice if it were easier to extract a grain's data in a form
that doesn't assume you're just going to re-create the grain. But maybe
we should treat that as a separate issue (do we have something
appropriate open already?) and use this to track the data corruption
problem with the existing feature.

Quoting Lauri Ojansivu (2020-01-18 09:25:19)

**Comment by xet7** [comment]
@zenhack 

Yes there is already separate issue #1324 , I'll add my comment there.

**Comment by kpreid** [comment]
Copying a grain's files while it is running is certainly hazardous, but **anything other than a copy of a grain's data has a significant risk that it might be _incomplete,_** either accidentally (bug in app) or consciously (export format cannot represent some data because it is an interchange format, or the app was designed assuming the export function was for a heavyweight OK-to-be-disruptive migration process); or the app might not even have an 'export' function (and requiring every app to have one feels against the spirit of Sandstorm freeing apps from needing administrative functions). Therefore, Sandstorm should continue to support the current 'orthogonal' form of backups.

Here's some possible options for reliable backups that do not involve app-specific code:

1. Snapshot grain while running (status quo). Problematically non-atomic, but cheap. Some apps could **declare that they are robust** against this backup strategy (e.g. trivial case: app has one data file which it updates by atomic rename).
2. Store grain data in a filesystem that supports atomic snapshots. This means the risk of corrupted backups is the same as the risk of corrupted grain data due to system crash/power off.
3. Backup operation shuts down grain. This is sound and straightforward, but might result in discarding the user's latest input in case the app is not good at writing changes promptly, and could be disruptive if the app has non-persistent session state so restarting the grain and reloading the page isn't sufficient.
4. Backups are taken automatically when the grain is not running; the user can choose to download a backup from such a time. This could also serve as part of a grain-level undo/version control feature.

**Comment by orblivion** [comment]
Oh, I like option 4. May be the best. In fact, users may even want to have
a trail of old snapshots for smaller grains.

I also had a crazy idea that I didn't want to bother mentioning, but what
the heck. It could spark a more realistic idea:

What if the backup process was to instantly fork the grain, both processes
and filesystem (requires something like UnionFS). The original one keeps
running as always. The fork has no network capability (in case it was in
the middle of making consequential outbound connections) and immediately
shuts down. After successful shutdown, it makes the zip.

**Comment by orblivion** [comment]
I also appreciate the plus side of options that don't require the app
developers create the backup process. It's in line with the rest of
Sandstorm's ease of safe development.

**Comment by zenhack** [comment]
I'm definitely in agreement that we want a standard backup facility that
doesn't rely on apps not to screw it up, and the current solution mostly
does that (though it could use some usability improvements) modulo this
bug.

I like some of these ideas. General thoughts:

* I'm not a fan of the idea of allowing apps to declare themselves
  robust to concurrent backups. I'd like to have a guarantee from the
  backup mechanism that an app can't mess up.
* Just shutting down the grain has the big advantage of simplicity.
  - If we're worried about the case where the user does something and
    then immediately hits "backup", we could wait a second or two before
    killing the grain, to increase the chances things have settled.
    Ultimately, apps have to be robust to sudden shutdowns, and we
    SIGKILL them on a regular basis so bugs are likely to be caught,
    so giving it a second or two should cover most cases; apps shouldn't
    be waiting too long before persisting data.
* Using filesystem snapshots is nice in that it gets us atomicity
  without being at all disruptive. The disadvantage is it imposes
  more constraints on the underlying system. In particular, this
  only works if you're installing on a Linux box that has a
  snapshot-capable filesystem, so we go from "run this script on some
  Linux box" to "make sure you're running btrfs or such (you're probably
  not, so ...reinstall???), then run this script." I don't like that.
* I like the idea of doing regular grain backups during downtime. We
  really should be automating backup processes as much as possible.
  - If we're doing this anyway, imposing the "shut down the grain before
    backing up" rule shouldn't be a big deal, since doing it manually
    won't be the common case.

**Comment by kpreid** [comment]
> we could wait a second or two before killing the grain,

Eek, a one-size-fits-all timeout! And, one that sounds a bit short for possible browser-server sync intervals (though perhaps we don't particularly want to support "user types something and then hits the backup button really fast").

How about, supposing this approach, a protocol by which the grain can be asked to save and respond when it has? (Of course, that's another point of failure, and raises the question of whether it should be used on all grain shutdowns…)

> filesystem snapshots … imposes more constraints on the underlying system

A possibility to avoid adding constraints would be to provide a snapshotting filesystem which is a proxy for the existing filesystem. I don't know whether something like this exists already for Sandstorm to use, but it seems like it could have uses beyond Sandstorm if it did come to exist.

> We really should be automating backup processes as much as possible.

👍

**Comment by zenhack** [comment]
Quoting Kevin Reid (2020-01-19 00:41:15)

> Eek, a one-size-fits-all timeout! And, one that sounds a bit short for
> possible browser-server sync intervals (though perhaps we don't
> particularly want to support "user types something and then hits the
> backup button really fast").

Fwiw, the existing grain shutdown mechanism also has a one-size-fits-all
timeout before shutdown, albeit a longer one.

But honestly I'm not sure it's even worth the trouble of offering a
grace period. My inclination is that the user generally won't back up
the grain while it is visibly in the middle of some big modification
doing work for them, and it isn't the norm for web apps to delay writing
persistent state for long.

> How about, supposing this approach, a protocol by which the grain can
> be asked to save and respond when it has? (Of course, that's another
> point of failure, and raises the question of whether it should be used
> on all grain shutdowns)

I remember this being discussed on the mailing list at length a few
years back in the context of regular grain shutdowns. We ultimately
rejected the idea, and the fact that we give grains no notice is very
deliberate and purposeful: fundamentally, we cannot save them from
unclean shutdowns; power outages happen, that's life. Providing a
notification with a grace period just makes sure that true failure
recovery never gets tested. I am against giving the app explicit notice
for backup shutdowns, for the same reason.

>      filesystem snapshots � imposes more constraints on the underlying
>      system
>
>    A possibility to avoid adding constraints would be to provide a
>    snapshotting filesystem which is a proxy for the existing filesystem. I
>    don't know whether something like this exists already for Sandstorm to
>    use, but it seems like it could have uses beyond Sandstorm if it did
>    come to exist.

IIRC blackrock creates a disk image for each grain (formatted with ext4)
rather than storing them directly on the host's filesystem. I think in
this case the point was so that nodes could connect to a grain's storage
via nbd. We could do something similar with btrfs images, so we don't
have to rely on details of the host filesystem.

-Ian

**Comment by xet7** [comment]
Quoting @zenhack

> apps shouldn't be waiting too long before persisting data.

> My inclination is that the user generally won't back up
the grain while it is visibly in the middle of some big modification
doing work for them, and it isn't the norm for web apps to delay writing
persistent state for long.

Meteor has [Optimistic UI](https://blog.meteor.com/optimistic-ui-with-meteor-67b5a78c3fcf), so UI changes are made visible to user immediately, and
saving to database happens later, UI change is only reverted if saving to database does not work.
Also [Apollo Starter Pack has Optimistic UI](https://github.com/sysgears/apollo-universal-starter-kit/blob/master/docs/Features%20and%20Modules.md#optimistic-ui-updates).

Because of [excessive server CPU usage](https://github.com/wekan/wekan/issues/718#issuecomment-350558222), [mixmax:smart-disconnect](https://atmospherejs.com/mixmax/smart-disconnect) was
[originally added to Wekan](https://github.com/wekan/wekan/blob/master/CHANGELOG.md#v061-2017-12-12-wekan-release) so that when tab is not used, connection to server was stopped to save server resources.

I did [remove mixmax:smart-disconnect](https://github.com/wekan/wekan/blob/master/CHANGELOG.md#v303-2019-08-04-wekan-release), because:
1. Users did choose another tab, so changes were not saved, and then close webbrowser, so there was data loss.
2. Users are working on multiple boards at different browser tabs and switching all time time, there was constant loading, so now after removing all tabs keeps active.

So, there is still high server CPU usage. This is also because Wekan has some subscriptions/publications that push data all the time to browser, like [Wekan bug that Linked cards make load all cards of database](https://github.com/wekan/wekan/issues/2688), so that browser caches become full of data and slow down to crawl.

If I remember reading correctly, in (current or newest) Meteor it's possible to:
a) Create Meteor project without minimongo.
b) Not use subscriptions/publications at all. Only use Meteor methods to call serverside, in that case frontend would behave similar to calling a REST API. That would probably need more code, and I don't know would database changes be immediately visible at frontend anymore.
c) Change big subscriptions to use smaller subscriptions, like subscribing to only _id.
d) Getting [Infinite scroll](https://github.com/wekan/wekan/pulls?q=is%3Apr+infinite+scroll+is%3Aclosed) working on all parts of Wekan. There is currently bug that [History of Comments only shows 20 newest comments max](https://github.com/wekan/wekan/issues/2377) so Infinite scroll does not work there.

But I have not yet tried which way of the above would work, and which way would be the best. Currently some Wekan users complain that Wekan server CPU usage is so high that loading some boards takes a lot of time, and scrolling big board with about 40 swimlanes and no images is very slow. They would like that I would fix all those performance problems permanently so they wound not need to change to other Kanban software, because they like using Wekan very much.

I would like to know if there is some solutions to above Wekan performance problems when using Meteor. Yes I have also looked at some other web frameworks than Meteor, but changing framework is a lot of work.

On other web apps like [CryptPad](https://cryptpad.fr/), it was said at [FLOSS Weekly 532 Interview](https://twit.tv/shows/floss-weekly/episodes/532) that in CryptPad when data is not yet saved to database, there is blinking red dot (or something similar) to show while saving is still in progress.

I don't know how safe shutting down MongoDB immediately is, because some Wekan grains have previously become corrupted. Most crash resistant database currently is SQLite.

**Comment by kentonv** [comment]
I haven't read this whole thread, but some comments:

Oasis actually did atomic backups, because it had full control of the underlying block device so could temporarily pause writes. Self-hosted Sandstorm, however, cannot inject itself at this layer, and cannot assume that the underlying filesystem supports snapshotting.

I think the best solution here may be to SIGSTOP all of the processes in the sandbox in order to pause them in time, then take the backup, then SIGSTART them. This way the backup is atomic, with no need to kill the app, and without the app needing to be aware of backups. I think in theory SIGSTOP/SIGSTART are observable by the app, but I think that's OK.

(If Sandstorm used cgroups, we could use the "cgroup freezer" to pause the app in a way that isn't observable, but cgroups are another thing that are difficult to rely on if you don't "own" the host system.)

**Comment by xet7** [comment]
At [SQLite corrupt page](https://www.sqlite.org/howtocorrupt.html) is mention that it would be best to use [SQLite Backup API](https://www.sqlite.org/backup.html).

**Comment by zenhack** [comment]
Are systems that don't support cgroups really that common? The feature went in quite some time ago.

Maybe we could default to cgroups and use sigstop as a fallback; it would be nice to use the more robust option where available.

**Comment by ocdtrekkie** [comment]
@zenhack How difficult would it be to enable backups from the grain list? Specifically, to utilize the existing manual backup option, but do it as a checkbox the grain and press a download button?

I have opened grains solely for the purpose of backing them up, and this seems silly for multiple reasons, including this issue.

**Comment by ocdtrekkie** [comment]
And following the topic of "can we backup a grain while it's not running"... could we come up with a "simple" automatic backup strategy that would let Sandstorm specify a backup storage location (maybe in something like #3204), and then upon each shutdown of the grain it would copy it to the backup location?

**Comment by zenhack** [comment]
I haven't dug into that part of the code yet, so I don't know how much work it would be, but I'm sure it's doable and I think it's a good idea.

I don't really think doing a backup of a grain on every shutdown is a good idea, partially because I think it's reasonably likely that the grain will be opened again soon, partially because of space concerns.

But possibly we could mitigate the space issue by using the kind of de-duplication algorithm used by perkeep, which is actually pretty good with disk images and things. Maybe build it on libasuran (since we've already got a bit of rust floating around anyway...).

We could maybe mitigate the re-opening issue by trying to do backups at times that sandstorm has observed are generally low-activity (probably stupid o'clock in the morning for most people). We may want that logic in sandstorm for automatically managing OS updates and reboots when we tackle a sandstorm "distro" at some point anyway.

I'd prefer to have one blessed dedup strategy, because I'd like sandstorm to just handle it for the user and given that I don't think there's much value in making it configurable and it would introduce complexity.

**Comment by ocdtrekkie** [comment]
So like, in my case, I mounted a NAS share to do the manual backup in the docs that requires shutting down Sandstorm. My NAS has it's own features for storing old copies of files. So I would be perfectly happy with overwriting grain backups, and leaving it on the implementer to decide how to manage copies. Even if Sandstorm exported a single overwriting copy of a grain each time it was closed, that would provide very close to perfect backup of every file you have in the case of "my drive failed" or "my VPS provider banned my account", and would be "better" than a nightly backup where I might've lost 8 hours of work.

People's on-prem, not-on-prem, cloud, etc. opinions on backups are going to vary wildly, and the amount of copies of grains people will want will also vary. Rather than build a complicated solution on Sandstorm's end, I would rather we have a simple pipe that lets people use whatever backup strategy they are comfortable with for the data fed out. (At work, one of my pet peeves is having to defeat an application's built-in backup strategy so it stops filling hard drives with redundant data I don't want that my backup solution can't manage effectively.)

**Comment by zenhack** [comment]
I'm open to having alternative strategies, though I want to avoid burden of choice on the user and I want things to "just work" as much as possible.

**Comment by ocdtrekkie** [comment]
My impression is that overwriting grain backups and letting users decide if they want solutions that support more is going to be the most pain-free solution. (And it would also let us backup on grain shutdown without ballooning storage, which I think is a strong perk for keeping the backup gap narrow.)

Apple's Time Machine has set that kind of configuration as pretty well understood for backup storage devices of many flavors. As I said, I know my NAS can retain overwritten copies, as can my sync software, which I believe defaults to keeping overwritten copies in a hidden folder.

In a business environment, my backup software is going to backup the entire drive, and keep x full copies/incrementals, and let me browse a directory at various dates and times. (And in that case, my backups are then stored on a device which manages it's own deduplication strategy.)

I feel like we definitely need a strategy to get data out of the Sandstorm directory without shutting down Sandstorm, but outside of that, should stray away from building our own backup software.

**Comment by xet7** [comment]
Related talk, search for "mongodb" this page:
https://fosdem.org/2020/schedule/events/

There is this:
Percona Backup for MongoDB: Status and Plans
Open Source solution for consistent backups of multi-shard MongoDB

Slides:
https://fosdem.org/2020/schedule/event/perconamongodb/attachments/slides/3768/export/events/attachments/perconamongodb/slides/3768/Percona_Backup_for_MongoDB.pdf

Video webm:
https://video.fosdem.org/2020/UD2.119/perconamongodb.webm

Same Video mp4:
https://video.fosdem.org/2020/UD2.119/perconamongodb.mp4

**Comment by zenhack** [comment]
Just a note, I was looking at cgroups recently for unrelated reasons -- we really ought to be doing performance isolation of grains, it's pretty easy for a grain to DoS sandstorm right now -- and so if that ends up in sandstorm freezing a cgroup will be an easy extension.

Annoyingly, it does look like it's hard to rely on being able to actually access cgroups on a given host.

**Comment by zenhack** [comment]
Now that #3398 is merged, should we close this? The issue still theoretically exists if you're running Sandstorm on a system where cgroups2 isn't available (e.g. kernel is too old) or if Sandstorm isn't running as root (not generally the case for prod deployments). But I believe the issue should be solved for the vast majority of real prod deployments.

**Comment by ocdtrekkie** [comment]
My inclination is that if people are having this issue, they should transition to a platform that supports the grain freezing functionality.

I guess the big question for me is, should we warn admins in the UI of this? Especially once Sandstorm does automated backup management, people will expect it to "just work", so similar to the misconfigured host questions, we might want to warn people about the possibility of backup corruption on their platform.

**Comment by ocdtrekkie** [comment]
Okay, my proposed solution, considering that the requirements for #3398 are apparently only in the Linux kernel as of July 2019, and we'll likely see older kernels for many years to come:

- Merge #3435 to allow backup of grains from the grain list. Usually, these are not running.
- If cgroups.freeze is not available, the top bar "backup grain" button should show a message that backing up running grains is not supported on this server, and that the user should backup the grain from the grain list.
- If cgroups.freeze is not supported, when backing up from the grain list, kill the grain prior to taking a backup, in case the grain is running.
- Possibly display a warning somewhere in the admin panel indicating certain grain management features may not be available on older kernels.

**Comment by zenhack** [comment]
I don't think we need to disallow the top-bar button: we can just kill the grain. For many/most apps it shouldn't have a bit difference in effect, it just might take a bit longer to start running again if we have to shut down the grain entirely.

**Comment by ocdtrekkie** [comment]
My thought was that the user experience of a grain shutdown while it's open isn't ideal. We should at least warn people it will restart the grain if applicable. Whereas the user isn't doing anything with the grain if they are at the grain list.
